### PR TITLE
Make README more accessible to novices

### PR DESCRIPTION
**Pull request by rlue**
This is an extensive rewrite of the README written as a follow-up to issue #1840. It has the following goals:

* give explanatory context for users who are new to _both_ RSpec and RSpec rails,
* name section headings with the questions they answer,
* make each section concise enough that a reader can comprehend the visual structure of the document on a quick scroll-through,
* replace "redundant" info (code blocks that demonstrate concepts readily available in the cucumber doc scenarios, or upgrade notes) with links to corresponding cucumber docs,
* update for latest version of RSpec Rails (specifically re: recommendations for controller specs and feature specs).

Notably, I've removed the section on rake tasks, and mostly removed the section on FactoryBot/Capybara. I've also added a link to betterspecs.org as an RSpec style guide.

Thoughts?

### CODE REVIEW

File: features/view_specs/view_spec.feature [hunk 1]
```diff
@@ -1,6 +1,55 @@
 Feature: view spec
 
-  View specs live in spec/views and render view templates in isolation.
+  View specs should be placed in `spec/views`,
+  and verify the content of view templates in isolation
+  (that is, without invoking a controller).
```
Comment (JonRowe) [inline features/view_specs/view_spec.feature]: Theres no verification that takes place.

File: features/view_specs/view_spec.feature [hunk 2]
```diff
@@ -1,6 +1,55 @@
...
+
+  expect(rendered).to match(/text/)
+  ```
+
+  1. Use the `assign` method to set instance variables in the view.
+     Technically, `@widget = Widget.new` would work too,
+     but RSpec doesn't officially support this pattern.
+     (It only works as a side effect of the fact that
+     view specs include `ActionView::TestCase` behavior.
+     Be aware that it may be made unavailable in the future.)
```
Comment (JonRowe) [inline features/view_specs/view_spec.feature]: Please delete from Technically onwards, we don't want to encourage that usage. This whole step would be better thought of as "setup", e.g.
```
1. Setup up required data for the view, e.g. using the `assign` method.
```

File: features/view_specs/view_spec.feature [hunk 3]
```diff
@@ -1,6 +1,55 @@
...
+
+  2. Use the `render` method to render the view.
+
+  3. Set expectations against the resulting `rendered` object.
```
Comment (JonRowe) [inline features/view_specs/view_spec.feature]: Don't enclose `rendered` in back ticks here, its not code. Maybe "rendered" if you want to distinguish it. Also it's not an object but a template. Its actually several objects and not one a developer can access.

File: features/view_specs/view_spec.feature [hunk 4]
```diff
@@ -1,6 +1,55 @@
...
+
+    ```ruby
+    render :template => "events/show", :layout => "layouts/application"
+    ```
+
+  * View specs expose a `controller` object,
+    which can be used to set expectations about the route (path + parameters)
+    to the view template being tested.
+    Some attributes of this object are based on
+    the name of the view itself—that is, in a spec for `events/index.html.erb`:
```
Comment (JonRowe) [inline features/view_specs/view_spec.feature]: Spacing is a bit weird here...

File: features/view_specs/view_spec.feature [hunk 5]
```diff
@@ -1,6 +1,55 @@
...
+
+  * To apply a layout to the view template being rendered,
+    be sure to specify both the template and layout explicitly:
+
+    ```ruby
+    render :template => "events/show", :layout => "layouts/application"
+    ```
+
+  * View specs expose a `controller` object,
+    which can be used to set expectations about the route (path + parameters)
```
Comment (JonRowe) [inline features/view_specs/view_spec.feature]: Its not intended to test the path or parameters, but to allow the view to use that information for various things.

File: features/matchers/have_http_status_matcher.feature [hunk 6]
```diff
@@ -31,7 +35,7 @@ Feature: `have_http_status` matcher
     When I run `rspec spec`
     Then the examples should all pass
 
-  Scenario: Checking a symbolic response code
+  Scenario: Checking a response code by name
```
Comment (JonRowe) [inline features/matchers/have_http_status_matcher.feature]: Symbol is important here, strings won't work, please revert.

File: features/matchers/have_http_status_matcher.feature [hunk 7]
```diff
@@ -56,7 +60,7 @@ Feature: `have_http_status` matcher
     When I run `rspec spec`
     Then the examples should all pass
 
-  Scenario: Checking a general response code
+  Scenario: Checking a response code by type
```
Comment (JonRowe) [inline features/matchers/have_http_status_matcher.feature]: Its not type, its more of a grouping..

File: features/matchers/have_http_status_matcher.feature [hunk 8]
```diff
@@ -1,12 +1,16 @@
...
+
+  * numeric
+  * name (defined in `Rack::Utils::SYMBOL_TO_STATUS_CODE`)
+  * type (`:success`, `:missing`, `:redirect`, or `:error`)
 
   The matcher works on any `response` object. It is available for use in
   [controller specs](../controller-specs), [request specs](../request-specs), and [feature specs](../feature-specs).
 
-  Scenario: Checking a numeric response code
+  Scenario: Checking a response code by number
```
Comment (JonRowe) [inline features/matchers/have_http_status_matcher.feature]: numeric response code reads better...

File: features/matchers/have_http_status_matcher.feature [hunk 9]
```diff
@@ -1,12 +1,16 @@
 Feature: `have_http_status` matcher
 
   The `have_http_status` matcher is used to specify that a response returns a
-  desired status code.
+  desired status code. It accepts one argument in any of the following formats:
+
+  * numeric
+  * name (defined in `Rack::Utils::SYMBOL_TO_STATUS_CODE`)
+  * type (`:success`, `:missing`, `:redirect`, or `:error`)
```
Comment (JonRowe) [inline features/matchers/have_http_status_matcher.feature]: Rails name rather than type...

File: README.md [hunk 10]
```diff
@@ -1,602 +1,387 @@
...
-* [https://github.com/rspec/rspec-core](https://github.com/rspec/rspec-core)
-* [https://github.com/rspec/rspec-expectations](https://github.com/rspec/rspec-expectations)
-* [https://github.com/rspec/rspec-mocks](https://github.com/rspec/rspec-mocks)
+  [additional configuration is required]: http://rubydoc.info/gems/rspec-rails/file/Capybara.md
 
-## Feature Requests & Bugs
+### RSpec style guides
 
-See <http://github.com/rspec/rspec-rails/issues>
+* [Better Specs](http://www.betterspecs.org/)
```
Comment (JonRowe) [inline README.md]: Better specs are not recommended by us and contain many antipatterns, I will not merge a PR with this in.

File: README.md [hunk 11]
```diff
@@ -1,602 +1,387 @@
...
-    user = User.create!(:username => "jdoe", :password => "secret")
-    get "/login"
-    assert_select "form.login" do
-      assert_select "input[name=?]", "username"
-      assert_select "input[name=?]", "password"
-      assert_select "input[type=?]", "submit"
+RSpec.describe Member do  #
+  context 'in death' do   # (almost) plain English
+    it 'has a name' do    #
+      expect(dead_member.name).to eq('Robert Paulson')  # test code
```
Comment (JonRowe) [inline README.md]: Two things, please don't use constant names in describe blocks, and please don't use a persons name.

File: README.md [hunk 12]
```diff
@@ -1,602 +1,387 @@
...
-Note that because view specs mix in `ActionView::TestCase` behavior, any
-instance variables you set will be transparently propagated into your views
-(similar to how instance variables you set in controller actions are made
-available in views). For example:
+At the risk of grossly oversimplifying,
+there are two kinds of tests you can write:
+those that verify
+**an individual component of your application** (unit tests),
+and those that verify
+**how multiple parts of your application work as a whole** (integration tests).
```
Comment (JonRowe) [inline README.md]: Can you remove this whole paragraph, it does vastly over simplify things, none of these are unit tests as they test integration with Rails.

File: README.md [hunk 13]
```diff
@@ -1,602 +1,387 @@
...
 
-```ruby
-@widget = Widget.new
-render # @widget is available inside the view
-```
+RSpec Rails defines ten different _types_ of specs (7 unit, 3 integration)
```
Comment (JonRowe) [inline README.md]: Remove `(7 unit, 3 integration)` its confusing because all Rails style specs are integration specs formally, and only a couple are end to end.

File: README.md [hunk 14]
```diff
@@ -1,602 +1,387 @@
...
+for testing different parts of a typical Rails application.
+Many inherit from one of Rails’ built-in `TestCase` classes,
```
Comment (JonRowe) [inline README.md]: They all inherit from Rails built in tests in some way

File: README.md [hunk 15]
```diff
@@ -1,602 +1,387 @@
...
+
+- **Integration**
+
+  | Spec type       | Corresponding Rails test class        |
+  | --------------- | ------------------------------------- |
+  | [request][]     | [`ActionDispatch::IntegrationTest`][] |
+  | [feature][]\*   |                                       |
+  | [system][]      | [`ActionDispatch::SystemTestCase`][]  |
+
+_\* discouraged in favor new alternatives_
```
Comment (JonRowe) [inline README.md]: Feature is not discouraged.

File: README.md [hunk 16]
```diff
@@ -1,602 +1,387 @@
...
+  | [job][]          |                                    |
+  | [view][]         | `ActionView::TestCase`             |
+  | [routing][]      |                                    |
+  | [helper][]       | `ActionView::TestCase`             |
+
+- **Integration**
+
+  | Spec type       | Corresponding Rails test class        |
+  | --------------- | ------------------------------------- |
+  | [request][]     | [`ActionDispatch::IntegrationTest`][] |
```
Comment (JonRowe) [inline README.md]: These are almost identical to controller tests and replace them (see earlier point about why referring to things as unit tests is unhelpful)

File: README.md [hunk 17]
```diff
@@ -1,602 +1,387 @@
...
+[model]: https://www.relishapp.com/rspec/rspec-rails/docs/model-specs
+[controller]: https://www.relishapp.com/rspec/rspec-rails/docs/controller-specs
+[mailer]: https://relishapp.com/rspec/rspec-rails/docs/mailer-specs
+[job]: https://relishapp.com/rspec/rspec-rails/docs/job-specs/job-spec
+[view]: https://www.relishapp.com/rspec/rspec-rails/docs/view-specs/view-spec
+[routing]: https://www.relishapp.com/rspec/rspec-rails/docs/routing-specs
+[helper]: https://www.relishapp.com/rspec/rspec-rails/docs/helper-specs/helper-spec
+[`ActionDispatch::IntegrationTest`]: https://edgeapi.rubyonrails.org/classes/ActionDispatch/IntegrationTest.html
+[`ActionDispatch::SystemTestCase`]: https://edgeapi.rubyonrails.org/classes/ActionDispatch/SystemTestCase.html
+[`ActionController::TestCase`]: https://edgeapi.rubyonrails.org/classes/ActionController/TestCase.html
```
Comment (JonRowe) [inline README.md]: Don't link to edge api please.

File: README.md [hunk 18]
```diff
@@ -1,602 +1,387 @@
...
+
+### I’m writing a spec file from scratch. How do I assign it a type?
+
+Assuming you haven’t [modified the default `rails_helper.rb` configuration][],
+simply place the spec in the appropriate folder
+(_e.g.,_ `spec/models/` for model specs)
+and RSpec will set its type automatically.
```
Comment (JonRowe) [inline README.md]: Don't mention this, we don't want to encourage it.

File: README.md [hunk 19]
```diff
@@ -1,602 +1,387 @@
...
 
-```ruby
-render
-expect(rendered).to match /Some text expected to appear on the page/
-```
-
-#### Upgrade note
+RSpec Rails provides three types of specs
+for integration testing the application as a whole—in other words,
+specifying what the client sees when interacting with it.
```
Comment (JonRowe) [inline README.md]: Not technically true, only feature / system specs have any javascript capability so they're the only ones capable of being end to end. Also integration tests don't have to be end to end, see earlier notes about confusing terminology.

File: README.md [hunk 20]
```diff
@@ -1,602 +1,387 @@
...
-Helper specs default to residing in the `spec/helpers` folder. Tagging any
-context with the metadata `:type => :helper` treats its examples as helper
-specs.
-
-Helper specs mix in ActionView::TestCase::Behavior. A `helper` object is
-provided which mixes in the helper module being spec'd, along with
-`ApplicationHelper` (if present).
+If you don’t have the luxury of upgrading,
+you’ll be limited to writing feature specs for your browser tests.
+They require [Capybara][],
```
Comment (JonRowe) [inline README.md]: So do system specs, ... its not limiting, it just predates Rails setup for it.

File: README.md [hunk 21]
```diff
@@ -1,602 +1,387 @@
...
+so they’re faster than feature specs,
+but do not examine your app’s UI or JavaScript.
 
-```ruby
-expect(:get => "/widgets").to route_to(:controller => "widgets", :action => "index")
-```
+Request specs provide a high-level alternative to unit testing controllers.
+In fact, as of RSpec 3.5,
+the RSpec team [discourages controller specs entirely][]
+in favor of request specs.
```
Comment (JonRowe) [inline README.md]: The Rails team makes this recommendation also.

File: features/view_specs/view_spec.feature (hunk 1 above)
Comment (rlue) [reply features/view_specs/view_spec.feature]: Alright; is 'test' acceptable instead?

Comment (rlue) [reply features/view_specs/view_spec.feature]: As in:

> Use them to test the content of view templates in isolation (that is, without invoking a controller).

File: features/view_specs/view_spec.feature (hunk 2 above)
Comment (rlue) [reply features/view_specs/view_spec.feature]: Done; this is [in the current version of the README](https://github.com/rspec/rspec-rails#assignkey-val), so I tried to find another place for it instead of deleting it.

File: features/view_specs/view_spec.feature (hunk 3 above)
Comment (rlue) [reply features/view_specs/view_spec.feature]: TIL.

File: README.md (hunk 11 above)
Comment (rlue) [reply README.md]: Fixed–but when you generate spec files with `rails g`, isn't that how they're formatted?

File: README.md (hunk 12 above)
Comment (rlue) [reply README.md]: 100%.

File: README.md (hunk 13 above)
Comment (rlue) [reply README.md]: Done.

File: README.md (hunk 14 above)
Comment (rlue) [reply README.md]: Fixed.

File: README.md (hunk 15 above)
Comment (rlue) [reply README.md]: Stricken.

File: README.md (hunk 17 above)
Comment (rlue) [reply README.md]: Whoops, great catch.

File: README.md (hunk 16 above)
Comment (rlue) [reply README.md]: Good point; I've removed the unit/integration dichotomy from the document entirely.

File: README.md (hunk 18 above)
Comment (rlue) [reply README.md]: Got it.

File: README.md (hunk 19 above)
Comment (rlue) [reply README.md]: Understood.

How do you feel about the following distinction instead?

> RSpec Rails provides three types of specs that do not directly correspond to any Rails application component.

I think this was the source of my original confusion about what certain spec types were for, and is (for me) the main criterion that distinguishes which spec types are worth explaining in detail and which should simply be linked to the Cucumber docs.

File: README.md (hunk 20 above)
Comment (rlue) [reply README.md]: Fixed.

File: README.md (hunk 21 above)
Comment (rlue) [reply README.md]: Added.

**Comment by rlue** [comment]
Thanks for your super fast feedback, guys. Learning a lot here.

File: README.md (hunk 19 above)
Comment (JonRowe) [reply README.md]: How about
```
RSpec Rails also provides some end to end (entire application) testing capability, to specify the interaction with the client.
```

Comment (rlue) [reply README.md]: My understanding was that you didn't want request specs grouped under the end-to-end umbrella. Would you like me to extract the **Request Specs** subsection into its own, separate section, or are you okay with leaving them there?

File: features/matchers/have_http_status_matcher.feature (hunk 9 above)
Comment (rlue) [reply features/matchers/have_http_status_matcher.feature]: I'm going to consolidate the scenario-naming conversation with this one, since they address the same issue.

I went back to the source code for this matcher, and noticed two things:

1. The "Rails name" arguments are defined in the [`GenericStatus` class](https://github.com/rspec/rspec-rails/blob/master/lib/rspec/rails/matchers/have_http_status.rb#L239), in a [`type_codes`](https://github.com/rspec/rspec-rails/blob/master/lib/rspec/rails/matchers/have_http_status.rb#L315) method. Thus, I'm proposing the term "generic status type".
2. The four possible values for generic status types are defined explicitly in the aforementioned `type_codes` method. I can't find any other place that lists exactly those four possible values (the [`ActionDispatch::TestResponse` documentation](https://edgeapi.rubyonrails.org/classes/ActionDispatch/TestResponse.html) doesn't include `#redirect?`, and the `Rack::Response::Helpers` module that it mixes in [doesn't include `#missing?` and _does_ include many others](https://github.com/rack/rack/blob/master/lib/rack/response.rb#L113-L134)), so I'm proposing we just list them explicitly here.

```
* numeric code
* status name, as defined in `Rack::Utils::SYMBOL_TO_STATUS_CODE`
* generic status type (`:success`, `:missing`, `:redirect`, or `:error`)

Scenario: Checking a numeric status code
Scenario: Checking a symbolic status name
Scenario: Checking a symbolic generic status type
```

File: README.md [hunk 22]
```diff
@@ -1,602 +1,360 @@
...
-    it "orders them in reverse chronologically" do
-      post = Post.create!
-      comment1 = post.comments.create!(:body => "first comment")
-      comment2 = post.comments.create!(:body => "second comment")
-      expect(post.reload.comments).to eq([comment2, comment1])
-    end
-  end
-end
+# Run a single example from a spec file (by line number)
+$ bundle exec rspec spec/controllers/accounts_controller_spec.rb:8
```
[the remark this answers is not in the document]
Comment (rlue) [reply README.md]: Good point, I like the parallelism.

File: README.md [hunk 23]
```diff
@@ -1,602 +1,360 @@
...
-RSpec.describe "home page", :type => :request do
-  it "displays the user's username after successful login" do
-    user = User.create!(:username => "jdoe", :password => "secret")
-    get "/login"
-    assert_select "form.login" do
-      assert_select "input[name=?]", "username"
-      assert_select "input[name=?]", "password"
-      assert_select "input[type=?]", "submit"
+RSpec.describe 'Post' do           #
+  context 'before publication' do  # (almost) plain English
```
[the remark this answers is not in the document]
Comment (rlue) [reply README.md]: The original inspiration for this PR was [a question I had as a beginner](https://github.com/rspec/rspec-rails/issues/1840)—"what is the difference between a 'spec' and a 'test'?" My goal was to show that the plain-English aspect of the RSpec DSL is what makes them "specs" (while the code component makes them "tests"). Sometimes, that "plain English" is hard to see, especially when it's hidden across three different method calls (`describe` / `context` / `it`). Hence, the comment.

I think for the majority of beginners, RSpec Rails will be their first introduction to RSpec. I thought it made sense to give a _very_ brief introduction to how RSpec works, and then leave it up to the reader to dig deeper into RSpec documentation.

Just my opinion; I'm open to removing it if y'all think it's too much.

[the remark this answers is not in the document]
Comment (benoittgt) [reply README.md]: I am ok with this. This is only a comment.

File: README.md [hunk 24]
```diff
@@ -1,602 +1,361 @@
...
-* [https://github.com/rspec/rspec](https://github.com/rspec/rspec)
-* [https://github.com/rspec/rspec-core](https://github.com/rspec/rspec-core)
-* [https://github.com/rspec/rspec-expectations](https://github.com/rspec/rspec-expectations)
-* [https://github.com/rspec/rspec-mocks](https://github.com/rspec/rspec-mocks)
+### Recommended third-party extensions
 
-## Feature Requests & Bugs
+* [FactoryBot](https://github.com/thoughtbot/factory_bot)
+* [Capybara](https://github.com/jnicklas/capybara)  
+  (Included by default in Rails 5.1+.
```
Comment (benoittgt) [inline README.md]: It should be on the same line as 356 no?

Comment (rlue) [reply README.md]: The newline serves two purposes:

1. to keep each line of text under 80 chars, and
2. to add a `<br>` between the Capybara link and the parenthetical note

That is, this markdown:

```markdown
* foo<space><space>
  bar
```

compiles to this HTML:

```html
<li>
  foo
  <br>
  bar
</li>
```

Originally, this was my preferred formatting, since the only other way to make a multi-line list item is with `\n\n`, which compiles to a `<p>` and results in a lot of vertical padding. But since now it looks like trailing spaces break the CI, I'm happy to just keep it all on one line instead.

File: README.md [hunk 25]
```diff
@@ -1,602 +1,361 @@
...
 
-Use **[rspec-rails 1.x](http://github.com/dchelimsky/rspec-rails)** for Rails
-2.x.
+`rspec-rails` brings the [RSpec][] testing framework to [Ruby on Rails][]
+as a drop-in alternative to its default testing framework, Minitest.
 
-## Installation
+In RSpec, tests are not just scripts that verify your application code;
+they’re also detailed explanations of how the application is supposed to behave,
+expressed in plain English—in other words, _specs._
```
Comment (JonRowe) [inline README.md]: Grammatically this bugs me, `English-in other words`, I think `English, or in other words,` might be better.

File: README.md [hunk 26]
```diff
@@ -1,602 +1,361 @@
...
+# rspec-rails [![Build Status][]][travis-ci] [![Code Climate][]][code-climate]
 
-Use **[rspec-rails 1.x](http://github.com/dchelimsky/rspec-rails)** for Rails
-2.x.
+`rspec-rails` brings the [RSpec][] testing framework to [Ruby on Rails][]
+as a drop-in alternative to its default testing framework, Minitest.
 
-## Installation
+In RSpec, tests are not just scripts that verify your application code;
+they’re also detailed explanations of how the application is supposed to behave,
```
Comment (JonRowe) [inline README.md]: Can we have `they are` apparently expanding contractions is better for non native speaker understandability.

File: README.md [hunk 27]
```diff
@@ -1,602 +1,361 @@
...
-Want to run against the `master` branch? You'll need to include the dependent
-RSpec repos as well. Add the following to your `Gemfile`:
+## Installation
 
-```ruby
-%w[rspec-core rspec-expectations rspec-mocks rspec-rails rspec-support].each do |lib|
-  gem lib, :git => "https://github.com/rspec/#{lib}.git", :branch => 'master'
-end
-```
+1. Add `rspec-rails` to **both** the `:development` and `:test` groups
```
Comment (JonRowe) [inline README.md]: Technically only the `:test` environment is required, but `:development` allows use of the generators, is it worth explaining that?

File: README.md (hunk 18 above)
Comment (JonRowe) [reply README.md]: I'd still like this changed to describe assigning a type by hand, and not mentioning setting type automatically.

File: README.md [hunk 28]
```diff
@@ -1,602 +1,361 @@
...
+to specify the interaction with the client.
 
-View specs default to residing in the `spec/views` folder. Tagging any context
-with the metadata `:type => :view` treats its examples as view specs.
+#### System specs
 
-View specs mix in `ActionView::TestCase::Behavior`.
+Also called **acceptance tests**, **browser tests**, or **end-to-end tests**,
+system specs test the application from the perspective of a _human client._
+The test code walks through a user’s browser interactions,
```
Comment (JonRowe) [inline README.md]: This should mention you could also use feature specs.

File: README.md [hunk 29]
```diff
@@ -1,602 +1,361 @@
...
-controller.controller_path == "events"
-controller.request.path_parameters[:controller] == "events"
-```
+#### Feature specs
 
-This means that most of the time you don't need to set these values. When
-spec'ing a partial that is included across different controllers, you _may_
-need to override these values before rendering the view.
+Feature specs serve the same purpose,
+but `SystemTestCase` solves some longstanding configuration issues they had,
```
Comment (JonRowe) [inline README.md]: It doesn't solve some longstanding configuration issues, it just has Rails handle it for you.

**Comment by benoittgt** [comment]
Thanks for the last changes. It's much better with them.

File: README.md (hunk 25 above)
Comment (JonRowe) [reply README.md]: No because `in other words, specs` is not a sentence, therefore a semicolon isn't correct either.

Comment (rlue) [reply README.md]: You're absolutely right. How would you feel about a colon there instead?

File: README.md (hunk 26 above)
Comment (JonRowe) [reply README.md]: @xaviershay would you mind pitching in here? I believe I learnt the "don't use contractions" rule from your good self.

File: README.md (hunk 25 above)
Comment (JonRowe) [reply README.md]: I'd prefer a normal comma.

Comment (rlue) [reply README.md]: Sure thing.

File: README.md [hunk 30]
```diff
@@ -1,602 +1,370 @@
...
+         create  .rspec
+         create  spec
+         create  spec/spec_helper.rb
+         create  spec/rails_helper.rb
 
-Use model specs to describe behavior of models (usually ActiveRecord-based) in
-the application.
+   # Rails 4+ only: generate a binstub
+   # (so you can run `bin/rspec` instead of `bundle exec rspec`)
+   $ bundle binstubs rspec-core
```
Comment (samphippen) [inline README.md]: @JonRowe I honestly *never* use binstubs, I don't know if we want this to be the default advice?

File: README.md [hunk 31]
```diff
@@ -1,602 +1,370 @@
...
-```
-
-### `rendered`
-
-This represents the rendered view.
-
-```ruby
-render
-expect(rendered).to match /Some text expected to appear on the page/
+RSpec.describe User, type: :model do
```
Comment (samphippen) [inline README.md]: could we bring this example up to above where the table of types is?

**Comment by samphippen** [comment]
@rlue I had a small question (for @JonRowe) and one requested change. Do let me know once you can get to them.

File: README.md (hunk 18 above)
Comment (JonRowe) [reply README.md]: This still needs inverting. The default should be writing types, that it can happen automatically is likely going to be removed in a later RSpec version, so it should not be mentioned.

File: README.md (hunk 30 above)
Comment (JonRowe) [reply README.md]: I don't use them, but Rails does, so I feel like maybe this advice should remain, happy to take you lead on this though

Comment (samphippen) [reply README.md]: I sort of want this to be in here, but de-emphasize it. Maybe pull it out in to a separate code box as an optional step?

Comment (JonRowe) [reply README.md]: Ok by me

Comment (samphippen) [reply README.md]: @rlue could you pull this out in to a separate code block and say that it's optional if you prefer binstubs?

**Comment by rlue** [comment]
Hey happy new year guys! Sorry for the delay; I've been out of the country for the past two weeks.

@samphippen, what do you think of this round of changes?

File: README.md (hunk 18 above)
Comment (rlue) [reply README.md]: Sorry @JonRowe, just noticed this now. Does the current version work for you?

File: README.md (hunk 30 above)
Comment (rlue) [reply README.md]: I actually went so far as to move it to the Usage section. My reasoning here is that before you've seen examples of `bundle exec rspec` usage, it may be difficult to grasp why you might even want a binstub in the first place.

As an ancillary benefit, it distills the Installation section down to the barest minimum.

File: README.md (hunk 31 above)
Comment (rlue) [reply README.md]: Like [this](https://github.com/rspec/rspec-rails/pull/2042/files#diff-04c6e90faac2675aa89e2176d2eec7d8R236)?

**Comment by samphippen** [comment]
This looks really good!

**Comment by samphippen** [comment]
@rlue if you rebase this against master, you should have your build fixed because of the bundler issue.

File: README.md [hunk 32]
```diff
@@ -1,602 +1,367 @@
...
 
-Use **[rspec-rails 1.x](http://github.com/dchelimsky/rspec-rails)** for Rails
-2.x.
+`rspec-rails` brings the [RSpec][] testing framework to [Ruby on Rails][]
+as a drop-in alternative to its default testing framework, Minitest.
 
-## Installation
+In RSpec, tests are not just scripts that verify your application code.
+They’re also detailed explanations of how the application is supposed to behave,
+expressed in plain English, or in other words, _specs._
```
Comment (mikegee) [inline README.md]: Consider including the word "specification" here. Readers may not know about the abbreviation.

Comment (rlue) [reply README.md]: Good point. Fixed!

**Comment by rlue** [comment]
@samphippen, my understanding is that rebasing rewrites history, which would make it impossible to push to the same remote branch, so I just merged instead—is that okay?

@JonRowe, I believe the only remaining open critique you had was about the contraction (they're) on line 7. Let me know if I'm missing anything else, or if that's a dealbreaker for you.

**Comment by rlue** [comment]
Thanks for everything, guys. The Travis CI build is still failing, and I'm puzzling over why.

The build that's failing is for 42bcf1b, a merge commit of fe43309 (master, [passing CI](https://travis-ci.org/rspec/rspec-rails/builds/473869488)) and 368e747 (from my fork, [failing CI](https://travis-ci.org/rspec/rspec-rails/builds/473869488)). A diff of 42bcf1b / fe43309 shows that [the only changed files are `README.md` and a couple of cucumber features](https://github.com/rspec/rspec-rails/compare/fe43309ac04af500ab666b28e9fae5777ebf5310...42bcf1b9aa41122312cf6f0227bc60d6df354c3e). The CI failures are all related to Bundler.

Some look like this:

```
$ rvm use 2.2.10 --install --binary --fuzzy
...
ruby-2.2.10 - #importing gemset /home/travis/.rvm/gemsets/global.gems..................................there was an error installing gem bundler
...
$ gem install bundler
ERROR:  Error installing bundler:
	The last version of bundler (>= 0) to support your Ruby & RubyGems was 1.17.3. Try installing it with `gem install bundler -v 1.17.3`
	bundler requires Ruby version >= 2.3.0. The current ruby version is 2.2.0.
The command "gem install bundler" failed and exited with 1 during .
```

Others look like this:

```
$ bundle install --binstubs --path ../bundle --retry=3 --jobs=3
The dependency jruby-openssl (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for java. To add those platforms to the bundle, run `bundle lock --add-platform java`.
Fetching gem metadata from https://rubygems.org/........
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Bundler could not find compatible versions for gem "bundler":
  In Gemfile:
    rails (~> 4.2.0) was resolved to 4.2.11, which depends on
      bundler (>= 1.3.0, < 2.0)
  Current Bundler version:
    bundler (2.0.1)
This Gemfile requires a different version of Bundler.
Perhaps you need to update Bundler by running `gem install bundler`?
Could not find gem 'bundler (>= 1.3.0, < 2.0)', which is required by gem 'rails
(~> 4.2.0)', in any of the sources.
Bundler could not find compatible versions for gem "cucumber":
  In Gemfile:
    cucumber (~> 1.3.5)
    aruba (~> 0.5.4) was resolved to 0.5.4, which depends on
      cucumber (>= 1.1.1)
Bundler could not find compatible versions for gem "nokogiri":
  In Gemfile:
    nokogiri (= 1.8.5)
    capybara (~> 2.2.0) was resolved to 2.2.1, which depends on
      nokogiri (>= 1.3.3)
The command "eval bundle install --binstubs --path ../bundle --retry=3 --jobs=3 " failed. Retrying, 2 of 3.
```

I haven't looked at every single failing test case (there are a lot), but all the ones I've seen look like one of the two above.

Any ideas?

**Comment by JonRowe** [comment]
Build is broken pending #2061

**Comment by JonRowe** [comment]
#2061 is merged so a rebase/merge would get this green hopefully.

**Comment by rlue** [comment]
All ready!

**Comment by samphippen** [comment]
Thank you for all your work here @rlue

**Comment by rlue** [comment]
Same to all three of you (especially @JonRowe)!
